tutorials/5_client_data.md

```diff
@@ -0,0 +1,486 @@
+which give invariants over fixed size subsets of clients. For example, we could
+specify that the investments of any two clients are conserved across calls to
+`Fund.transfer(address, uint256)`. This property will be our running example for
+the remainder of the tutorial. We specify the property in past linear temporal
+logic (pLTL) by writing:
+
+> It is *always* the case that whenever `msg.sender` calls `Fund.transfer()` and
+> sends `_amount` to some other client `_dst`, then `Fund.invested[msg.sender]`
+> is decreased by  `_amount` while  `Fund.invested[_destination]` is increased
+> by `_amount`.
+
+We then formalize the property using the
+[VerX Specification Language](https://verx.ch/docs/spec.html), As presented in
+[tutorial 3](3_transactions.md). Recall that `FUNCTION` is the name of the last
+method called, while `prev(v)` is the value of `v` before `FUNCTION` was called.
+`once` and `always` are pLTL operators, where `once(p)` is true if `p` has ever
+been true while `always(p)` is true if `p` has always been true. We write
+`Fund.transfer(address, uint256)[i]` to refer to the i-th argument passed to
+`Fund.transfer(address, uint256)`. This gives the formalization:
+
+```
+always(
+    (
+        (FUNCTION == Fund.transfer)
+        &&
+        (msg.sender != Fund.transfer(address,uint256)[0])
+    )
+    ==>
+    (
+      (Fund.invested[msg.sender]
+            + Fund.transfer(address,uint256)[1]
+            == prev(Fund.invested[msg.sender]))
+        &&
+        (Fund.invested[Fund.transfer(address,uint256)[0]]
+            - Fund.transfer(address,uint256)[1]
+            == prev(Fund.invested[msg.sender]))
+    )
+)
+```
+
+As a final step, we can constructor a monitor for the property. The monitor
+detects if the property is violated, and allows us to verify the claim. Our
+property is essentially a functional post-condition, so the monitor is very
+compact. First we introduce the following predicates:
+
+  * `is_transfer := FUNCTION == Fund.transfer`
+  * `distinct := msg.sender != _destination`
+  * `sent := invested[msg.sender] + _amount = prev(invested[msg.sender])`
+  * `recv := prev(invested[_destination]) + _amount = prev(invested[_destination]`
+
+And then give the regular expression
+`(!(is_transfer && distinct) || (send && recv))*`.
+
+## Local Reasoning Over Client Mappings
+
+Mappings allow us to associated variables with clients. This fundamentally
+changes the `Manager` bundle topology. We can think of this new topology as a
+graph with two types of vertices: *process vertices* and *data vertices*
+[[2](#reference)]. Each process vertex is assigned to one or more transactions
+in the `Manager` bundle, whereas each data vertex is assigned to a single
+mapping entry. If a process can write to a mapping entry, there is a directed
+edge from the process vertex to the data vertex. If a process can read from a
+mapping vertex, there is a directed edge from the data vertex to the process
+vertex.
```

Comment:
Instead of assuming everyone knows and remembers what *topology* means, describe the graph that you want. Then say that this is called **topology** and refer to local reasoning paper.

Reply:
also good to have a picture